actionview/lib/action_view/lookup_context.rb

```diff
@@ -55,21 +61,50 @@ module Accessors #:nodoc:
     register_detail(:handlers) { Template::Handlers.extensions }
 
     class DetailsKey #:nodoc:
+      class SmallCache < Concurrent::Map
+        def initialize(options = {})
+          super(options.merge(initial_capacity: 2))
+        end
+      end
+
+      # preallocate all the default blocks for performance/memory consumption reasons
+      PARTIAL_BLOCK = lambda { |cache, partial| cache[partial] = SmallCache.new }
+      PREFIX_BLOCK  = lambda { |cache, prefix|  cache[prefix]  = SmallCache.new(&PARTIAL_BLOCK) }
+      NAME_BLOCK    = lambda { |cache, name|    cache[name]    = SmallCache.new(&PREFIX_BLOCK) }
+      RESOLVER_BLOCK = lambda { |cache, resolver| cache[resolver] = SmallCache.new(&NAME_BLOCK) }
+      KEY_BLOCK     = lambda { |cache, key|     cache[key]     = SmallCache.new(&RESOLVER_BLOCK) }
+
       alias :eql? :equal?
 
       @details_keys = Concurrent::Map.new
+      @digest_cache = Concurrent::Map.new
+      @view_template_cache = SmallCache.new(&KEY_BLOCK)
 
-      def self.get(details)
+      def self.digest_cache(details)
         if details[:formats]
           details = details.dup
           details[:formats] &= Template::Types.symbols
         end
-        @details_keys[details] ||= Concurrent::Map.new
+        @digest_cache[details] ||= Concurrent::Map.new
+      end
+
+      def self.details_cache_key(details)
+        if details[:formats]
+          details = details.dup
+          details[:formats] &= Template::Types.symbols
+        end
+        @details_keys[details] ||= Object.new
```

Comment:
It is not clear to my why the cache key is an instance of object. It is just to create a unique value?

Reply:
Yes.  I think the original reason was that calculating a hash for the details key was expensive (essentially `{ some: :complex, thing: { a: b: } }.hash`), so we do it once and use an object instead.  Apparently we have to consult the template cache many times, so we need a fast `.hash`.

I think we can eliminate this cache all together and still maintain speed, but it may require changes to the lookup context.  That's why I'm interested in making the lookup context more private (or be more aggressive with API changes).

Reply:
Makes sense.